### ISSUE TITLE
Batch size when constructing dataset from HDF5

### ISSUE THREAD

**Issue by sh0gg0th** [microsoft/LightGBM#5361]
In the [multiple HDF5 example](https://github.com/microsoft/LightGBM/blob/master/examples/python-guide/dataset_from_multi_hdf5.py), we set a batch size of 64. Does the batch size influence the final LGBM dataset written to disk with `.save_binary` in any way, or is it relevant to HDF5 sequences only?

**Comment by jmoralez** [comment]
Hi @sh0gg0th, thank you for your interest in LightGBM. The dataset is constructed from a random sample of your full dataset https://github.com/microsoft/LightGBM/blob/273c9b043e11b24a95e40f99ef81c4888d2089c9/python-package/lightgbm/basic.py#L1623
By default the sample size is 20,000 https://github.com/microsoft/LightGBM/blob/273c9b043e11b24a95e40f99ef81c4888d2089c9/include/LightGBM/config.h#L623
Once the dataset is constructed all the rows are pushed in batches, which means that for each row, each feature gets assigned to its corresponding bin
https://github.com/microsoft/LightGBM/blob/273c9b043e11b24a95e40f99ef81c4888d2089c9/python-package/lightgbm/basic.py#L1625-L1630
This last part is where the batch size is used. It controls the size of the arrays that will be pushed in each batch, so it shouldn't affect the final dataset, just the memory used and the time it will take to create it.

Please let us know if you have further doubts.

**Comment by sh0gg0th** [comment]
Understood, thanks a lot for the detailed reply @jmoralez ! So there aren't really any manual optimizations to be done if I want to minimize disk reads during training and utilize my RAM to the fullest?

**Comment by jmoralez** [comment]
The HDF5 example allows you to create a [LightGBM Dataset](https://lightgbm.readthedocs.io/en/v3.3.2/pythonapi/lightgbm.Dataset.html#lightgbm-dataset) and save it in binary format, which is a compressed format that can be used by LightGBM directly. When you train, however, you'll need to load the full data set. The advantage is that this format is more compact and can probably fit on your machine. Here's an example:

```python
import os

import numpy as np
import lightgbm as lgb

X = np.random.rand(100_000, 100)
np.save("arr.npy", X)
print(f'Array size: {os.path.getsize("arr.npy") / 2**20:.1f} MB')
lgb.Dataset(X).construct().save_binary("ds.bin")
print(f'Dataset size: {os.path.getsize("ds.bin") / 2**20:.1f} MB')
ds = lgb.Dataset("ds.bin")
lgb.train({'num_leaves': 3, 'verbose': -1}, ds, num_boost_round=1)
```
Which prints:
```
Array size: 76.3 MB
[LightGBM] [Info] Saving data to binary file ds.bin
Dataset size: 10.1 MB
```
In this case the memory used to hold the data would just be 10MB instead of 76.

Please let us know if you have further doubts.
